File: 5 - BuscaBinariaCEP/BuscaBinCep.py
import struct
import sys

if len(sys.argv) != 2:
  print(f"Uso python {sys.argv[0]} [CEP]")
  quit()

registroCEP = struct.Struct("72s72s72s72s2s8s2s")
cepColumn = 5

def buscaBin(inicio, fim, file, cepProcurado):
  count = 0

  while inicio <= fim:
    meio = (inicio + fim) // 2
    file.seek(meio * registroCEP.size, 0)
    line = file.read(registroCEP.size)
    # line = file.readline()
    count += 1

    record = registroCEP.unpack(line)

    cep_just_read = record[cepColumn].decode('latin1')

    if cepProcurado == cep_just_read:
      return file.tell(), count
    elif cepProcurado > cep_just_read:
      inicio = meio + 1
    else:
      fim = meio - 1
  return -1

with open('../datasets/cep_ordenado.dat', 'rb') as f:
  f.seek(0, 2)
  tamanho_bytes = f.tell()
  tamanho_registros = tamanho_bytes // registroCEP.size
  
  inicio = 0
  fim = tamanho_registros - 1

  idx, count = buscaBin(inicio, fim, f, sys.argv[1])

  if idx < 0:
    print("Nao encontrou")
  else:
    # f.seek(-1 * registroCEP.size, 1)
    # a partir da posição atual, volta 1 registro para que a proxima leitura seja o cep certo

    f.seek(idx - (registroCEP.size * 1), 0) 
    # a partir do inicio do arquivo, define a posição como sendo a posição do registro anterior
    print(f.readline().decode('latin1'))

  print(f"Enderecos lidos: {count}");

File: 5 - BuscaBinariaCEP/BuscaBinCep.c
#include <stdio.h>
#include <string.h>

typedef struct _Endereco Endereco;

// registroCEP = struct.Struct("72s72s72s72s2s8s2s")

struct _Endereco
{
	char logradouro[72];
	char bairro[72];
	char cidade[72];
	char uf[72];
	char sigla[2];
	char cep[8];
	char lixo[2]; // Ao Espaço no final da linha + quebra de linha
};

long buscaBin(int inicio, int fim, FILE *f, char* cepProcurado, Endereco* e, int* count){
    int meio;

    while(inicio <= fim){
        meio = (inicio + fim)/2;
        fseek(f, meio*sizeof(Endereco), SEEK_SET);
        fread(e, sizeof(Endereco), 1, f);
        *count += 1;
        if (strncmp(cepProcurado, e->cep, 8) == 0){
            return ftell(f);

        } else if(strncmp(cepProcurado, e->cep, 8) > 0){
            inicio = meio + 1;

        } else {
            fim = meio - 1;
        }
    }
    return -1;
}


int main(int argc, char**argv)
{
	FILE *f;
	Endereco e;
	int qt;
  long tamanho_bytes;
  long tamanho_registros;

	if(argc != 2)
	{
		fprintf(stderr, "USO: %s [CEP]", argv[0]);
		return 1;
	}

	printf("Tamanho da Estrutura: %ld\n\n", sizeof(Endereco));

	f = fopen("../datasets/cep_ordenado.dat","rb");

  fseek(f, 0, SEEK_END);
  tamanho_bytes = ftell(f);
  tamanho_registros = tamanho_bytes/sizeof(Endereco);

  int inicio = 0;
  int fim = tamanho_registros - 1;

  int count = 0;

  long idx = buscaBin(inicio, fim, f, argv[1], &e, &count);
  
  if(idx == -1){
      printf("Nao encontrou\n");
  } else {
    printf("%.72s\n%.72s\n%.72s\n%.72s\n%.2s\n%.8s\n",e.logradouro,e.bairro,e.cidade,e.uf,e.sigla,e.cep);
  }

  printf("Enderecos lidos: %d\n", count);
	fclose(f);
  return 0;
}